halo2_proofs/src/dev.rs

```diff
@@ -746,8 +746,32 @@ impl<F: FieldExt> MockProver<F> {
     /// Returns `Ok(())` if this `MockProver` is satisfied, or a list of errors indicating
     /// the reasons that the circuit is not satisfied.
     pub fn verify(&self) -> Result<(), Vec<VerifyFailure>> {
+        self.verify_at_rows(self.usable_rows.clone(), self.usable_rows.clone())
+    }
+
+    /// Returns `Ok(())` if this `MockProver` is satisfied, or a list of errors indicating
+    /// the reasons that the circuit is not satisfied.
+    /// Constraints are only checked at `gate_row_ids`,
+    /// and lookup inputs are only checked at `lookup_input_row_ids`
+    pub fn verify_at_rows<I: Clone + Iterator<Item = usize>>(
+        &self,
+        gate_row_ids: I,
+        lookup_input_row_ids: I,
+    ) -> Result<(), Vec<VerifyFailure>> {
         let n = self.n as i32;
 
+        // check all the row ids are valid
+        for row_id in gate_row_ids.clone() {
+            if !self.usable_rows.contains(&row_id) {
```

Comment:
Is it better to range check `row_id` against `usable_rows.end` and `usable_rows.start` rather than `contains`. As it is it could be time consuming to existence query for large `usable_rows`.

Reply:
Indeed, checking than the len of `gate_row_ids` exists in `self.usable_rows` should suffice if as @kilic said (and I agree) the vector contains sequential values.

Reply:
usable_rows here is Range not vector.   `contains` implementation of Range is indeed check low bound and high bound :  https://doc.rust-lang.org/src/core/ops/range.rs.html#810-813

Reply:
But the point is that it checks it for each `row_id`.
When indeed the only `row_ids` that matter are `gate_row_ids[0]` and `gate_row_ids[last_pos]`

Reply:
check row_id in Range is **much** cheaper than checking gates constraints, and checking only `gate_row_ids[0]` and `gate_row_ids[last_pos]`, we require caller to provided sorted row_ids.    I think the cost/benefit is better for checking every id?

Reply:
Ohhh I see now!! Sure. Makes sense now. Also with the improvements in testing time that you announced, we can merge this and check this minor things in the future.